iree/compiler/Dialect/Flow/Transforms/ReplaceFuncArgsWithConstants.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+namespace {
+
+// Replace the function arguments with constants if it has "iree.module.export"
+// attribute.
+class FuncArgsReplacementPattern : public OpRewritePattern<FuncOp> {
+  using OpRewritePattern<FuncOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(FuncOp funcOp,
+                                PatternRewriter &rewriter) const override {
+    if (!funcOp.getAttr("iree.module.export")) {
+      return failure();
+    }
+    if (funcOp.getNumArguments() == 0) {
+      return failure();
+    }
+    Location loc = funcOp.getLoc();
+    auto funcType = rewriter.getFunctionType({}, funcOp.getType().getResults());
+    auto newFuncOp = rewriter.create<FuncOp>(loc, funcOp.getName(), funcType);
+    newFuncOp.setAttr("iree.module.export", rewriter.getUnitAttr());
+    Block *block = newFuncOp.addEntryBlock();
+    auto blockBuilder = OpBuilder(block, block->begin());
+    BlockAndValueMapping mapping;
+    for (auto iter : llvm::enumerate(funcOp.getType().getInputs())) {
+      auto attr = blockBuilder.getZeroAttr(iter.value());
+      auto cst = blockBuilder.create<ConstantOp>(loc, iter.value(), attr);
```

Comment:
We want `iree.unfoldable_constant` here instread of `std.constant`

Reply:
We do want std.constant and rely on createOutlineLargeConstantsPass. This will move the constants to flow.variables; this will ensure that you get proper constant handling in the following HAL pipeline (vs this, which will be allocating and filling these buffers at runtime in the benchmark).

See #3431 for more details, this is in the same context there.

Reply:
SG! Thanks for explaining

Reply:
I'm a bit worried about making these constants look the same as normal constants (see point above). We *don't* want these to get proper constant handling in the sense that we don't want the compiler doing something clever because it knows this is a constant.

Reply:
Another way to do this that ensures that the constants are not messed with regardless of which passes follow is to create the `flow.variable` with the constant value and `noinline` attribute and replace the use of the constant with a `flow.variable.load`. That's exactly how SavedModels look when imported (besides the noinline, which is what prevents the constant folding).

Reply:
scott's new PR does something similar (creating the flow.variables) - you may be able to copy that: https://github.com/google/iree/pull/3629/files

Reply:
Sounds good, I create flow.variables myself and remove the outlining large constants pass.